### PR TITLE
[SPARK-31988][SQL] Schema pruning may discard attribute metadata

### PR DESCRIPTION
**Pull request by guykhazma**
### What changes were proposed in this pull request?

Fixing the `getRootFields` and `sortLeftFieldsByRight` functions to preserve attribute metadata

### Why are the changes needed?

This can lead to a potential loss of metadata on an attribute in some code paths.

For example - when reading a parquet file with a schema that has metadata and writing it back the parquet footer it will not contain the metadata that was originally there.

code snippet to reproduce (assuming datasource v2):
```Scala
    // create custom dataset
    val data = Seq(
      Row("a", "b")
    )
    val schema = List(
      StructField("col_a", StringType, true,
        new sql.types.MetadataBuilder().putString("key", "value").build()),
      StructField("col_b", StringType, true)
    )
    val df = spark.createDataFrame(
      spark.sparkContext.parallelize(data),
      StructType(schema)
    )
    // write
    df.write.parquet("/tmp/check")
    // read and verify the metadata exists
    val readDF = spark.read.parquet("/tmp/check")
    readDF.schema.foreach(s => println(s.metadata))
    // write again
    readDF.write.parquet("/tmp/check2")
    // read again and verify no metadata
    val readDF2 = spark.read.parquet("/tmp/check2")
    readDF2.schema.foreach(s => println(s.metadata))
```

Note that this doesn't happen in datasource v1.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

No tests were added as this is a minor change to a private function and no tests exist to check this function so far.

### CODE REVIEW

**Comment by maropu** [comment]
Could you add some tests first? Thanks!

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @viirya FYI

**Comment by guykhazma** [comment]
@maropu which test do you suggest to add?
This is private function which is not tested anywhere also not all file formats are able to save the metadata (for example csv) so seems to me that adding a test to [SchemaPruningSuite](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala) is not the right thing.
I can add a test similar to the above code snippet, if that seems ok to you where would you suggest to put it?

**Comment by maropu** [comment]
Basically, I think we need the two types of tests: end-2-end (coarse-grained) tets and fine-grained tests. For the former tests, how about adding some tests in `DataFrameSuite`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala [hunk 1]
```diff
@@ -120,7 +120,8 @@ object SchemaPruning {
   private def getRootFields(expr: Expression): Seq[RootField] = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:120]: How about `private` -> `private[expressions]` if there is no other way to add tests?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:120]: I think `getRootFields` is used for data source v1 and v2. Why only v2 is affected?

**Comment by dongjoon-hyun** [comment]
+1, for @maropu 's suggestion. You can use your example in the PR description, @guykhazma .

**Comment by dongjoon-hyun** [comment]
BTW, @guykhazma . Is the example in the PR description enough? If I follow the direction, the metadata is not lost.
```
scala> sc.version
res1: String = 3.0.0
...
{"key":"value"}
{}
{"key":"value"}
{}
```

**Comment by guykhazma** [comment]
@dongjoon-hyun did you make sure to run it with datasource v2?
I will add tests for both and a better explanation later today.

**Comment by guykhazma** [comment]
@HyukjinKwon @maropu @dongjoon-hyun @viirya added a test and also another fix that was needed - the function `sortLeftFieldsByRight` also ignored the `metadata` field.

As for why it happens only for v2 is because `V2ScanRelationPushDown` calls `pruneColumns` which trigger this code path (see [here](https://github.com/apache/spark/blob/d2a656c81ef784657a02e7347bfe87e4331fd2c9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala#L50))

**Comment by dongjoon-hyun** [comment]
Thanks for updating, @guykhazma .

**Comment by viirya** [comment]
> While v1 calls for `prunePhysicalColumns`
> 
> https://github.com/apache/spark/blob/d2a656c81ef784657a02e7347bfe87e4331fd2c9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala#L63
> 
> and the property `root.derivedFromAtt` is set to true so the structType from `requestedRootFields` is not used.

Thanks for the update, @guykhazma. So, does it mean if we have `derivedFromAtt` is false on any requested root field, v1 path will also discard attribute metadata. It sounds like we can also have a test case for that?

**Comment by guykhazma** [comment]
@viirya `derivedFromAtt` is set to `false` when the expression is requesting a nested field .
The metadata for a nested column is not preserved (also in Spark 2.4) so I am not sure what the expected behaviour should be here. 
Note that the metadata for nested field also not preserved when using:
```Scala
df.select("col_a.name").schema
```
when `df` is the dataframe that was created locally with the specified schema.
If this is considered a bug then we can resolve this as well (will require some more changes).

For example this test triggers a code path where `derivedFromAtt` is `false` but it currently passes since metadata is not preserved for nested columns:

```Scala
  test("SPARK-31988 - make sure schema metadata is preserved - nested schema") {
    withSQLConf((SQLConf.USE_V1_SOURCE_LIST.key, "avro,csv,json,kafka,orc,text,parquet")) {
      withTempPath{ f =>
        // create custom dataset with schema metadata
        val data = Seq(
          Row(Row("a", 45), "b")
        )
        val schema = List(
          StructField("col_a", StructType(
            List(
              StructField("name", StringType, true,
                new MetadataBuilder().putString("check", "b").build()),
              StructField("age", IntegerType, true)
            )
          ), true,
            new MetadataBuilder().putString("key", "value").build()),
          StructField("col_b", StringType, true)
        )

        val df = spark.createDataFrame(
          spark.sparkContext.parallelize(data),
          StructType(schema)
        )
        df.write.parquet(f.getAbsolutePath)

        // read from storage
        val readDF = spark.read.parquet(f.getAbsolutePath)
        // write again
        withTempPath { f =>
          readDF.select("col_a.name").write.parquet(f.getAbsolutePath)
          // read again and verify the schema is equal (including the metadata)
          val readDF2 = spark.read.parquet(f.getAbsolutePath)
          assert(readDF.select("col_a.name").schema == readDF2.schema)
        }
      }
    }
  }
```

**Comment by guykhazma** [comment]
@viirya @maropu any comments?

**Comment by guykhazma** [comment]
@viirya @maropu any comments?

**Pull request closed without merging**